data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3SelectObjectWorker.java

```diff
@@ -0,0 +1,218 @@
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+import org.apache.commons.io.input.CountingInputStream;
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.log.JacksonLog;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.source.configuration.S3SelectSerializationFormatOption;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.DistributionSummary;
+import software.amazon.awssdk.core.SdkClient;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.CompressionType;
+import software.amazon.awssdk.services.s3.model.ExpressionType;
+import software.amazon.awssdk.services.s3.model.FileHeaderInfo;
+import software.amazon.awssdk.services.s3.model.InputSerialization;
+import software.amazon.awssdk.services.s3.model.JSONOutput;
+import software.amazon.awssdk.services.s3.model.JSONType;
+import software.amazon.awssdk.services.s3.model.ParquetInput;
+import software.amazon.awssdk.services.s3.model.RecordsEvent;
+import software.amazon.awssdk.services.s3.model.SelectObjectContentEventStream;
+import software.amazon.awssdk.services.s3.model.SelectObjectContentRequest;
+
+/**
+ * Class responsible for taking an {@link S3SelectResponseHandler} and creating
+ * all the necessary {@link Event} objects in the Data Prepper {@link Buffer}.
+ */
+public class S3SelectObjectWorker implements S3ObjectHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(S3SelectObjectWorker.class);
+    static final String S3_OBJECTS_SIZE_PROCESSED = "s3ObjectProcessedBytes";
+    static final String S3_OBJECTS_FAILED_METRIC_NAME = "s3ObjectsFailed";
+    static final String S3_OBJECTS_SUCCEEDED_METRIC_NAME = "s3ObjectsSucceeded";
+    static final String S3_OBJECTS_EVENTS = "s3ObjectsEvents";
+    static final String S3_BUCKET_NAME = "bucket";
+    static final String S3_OBJECT_KEY = "key";
+    static final String S3_BUCKET_REFERENCE_NAME = "s3";
+    private S3AsyncClient s3AsyncClient;
+    private final DistributionSummary s3SelectSizeProcessedSummary;
+    private final DistributionSummary s3SelectEventsSummary;
+    private final Counter s3ObjectsSucceededCounter;
+    private final Counter s3SelectFailedCounter;
+    private Duration bufferTimeout;
+    private int numberOfRecordsToAccumulate;
+    private Buffer<Record<Event>> buffer;
+    private String query;
+    private String dataSerilizationFormat;
+
```

Comment:
It looks like some of these are repeated in S3ObjectWorker.java too. Is there a way to avoid duplicating this? It may be possible put the metrics initialization in the S3Service.java?

Reply:
Thanks, Krishna for your review. As per your comment this implementation require refactoring of existing code. if you want us to modify we need some more time to implement and test thoroughly. please suggest.